### DISCUSSION TITLE
How to disable concurrent login

### DISCUSSION THREAD

**Discussion by perlland** [ltsp/ltsp#357] [Q&A]
Is there any way to disable concurrent login in multiple clients?? Kindly help me to deny one single user to login to his account at multiple client machine same time....

**Comment by alkisg** [comment]
You can use limits.conf on the server to limit SSH(FS) connections to one. Example commands:

```shell
echo "@onelogin - maxlogins 1" > /etc/security/limits.d/local.conf
addgroup onelogin
gpasswd -a guest01 onelogin
```

These create a group called "onelogin", they impose a "maxlogins 1" limit to it, and they add a "guest01" user to that group. You're supposed to add all the restricted users to that group.

**Reply by perlland** [reply]
In my case only sudo account or primary account is restricted to concurrent-login... not for local user accounts.....Kindly help
![Screenshot from 2020-12-31 10-46-11](https://user-images.githubusercontent.com/36441218/103395708-858f5a80-4b55-11eb-8961-e7d33724f1ef.png)
In above screenshot, "work" is my primary account and user2/user3 are my client accounts....
Concurrent Login is working incase of account "work".. but not for the accounts "user2" and "user3"

Kindly help

**Comment by alkisg** [comment]
Unfortunately SSHFS connections don't count as "logins", so this only works if someone used SSH or directly connected to the server console.
We could use SSH instead of SSHFS, but it would consume more server resources, and more importantly, many LTSP users don't want to give SSH access to the server, they only want to give SSHFS access, possibly even in a chroot.
So one solution would be to use the "RequestTTY force" SSH option along with SSHFS. But there's an SSHFS bug preventing that; I reported it there, you can comment that it affects you:
https://github.com/osxfuse/sshfs/issues/63

Another solution would be to override the session with e.g. a /usr/local/bin/mate-session script. That would maintain a ~/.cache/ltsp/login file, it would prompt if it saw an existing login recorded there, and it could even monitor it in order to terminate the session if another login was started elsewhere. That way the following dialog could be implemented and used:

| Existing login detected |
|:--:|
| You are already logged in to the server from HOSTNAME=ltsp123, IP=192.168.67.123.<br>Do you want to terminate that session, and login from here?<br><br>`Yes` `Exit` `Login concurrently` |

Patches welcome for that solution (in python3/gtk, including plain shell support). Personally I've no plans to implement this as we have no need for it here.

**Reply by perlland** [reply]
How to know which **sshfs client-session is active** in LTSP Server?

**Comment by perlland** [comment]
How to know which sshfs client-session is active in LTSP Server? Can I use epoptes as a command to list, terminate activate sshfs client-session? Kindly help..

**Comment by alkisg** [comment]
You can run a command like `ps` and check for sftp-server processes.
But to know which ones of them are truly active, you'd need to re-read my previous answer above...

**Reply by perlland** [reply]
I can get process ID of Client-sftp connection and kill these. But killing those process alone disconnects client connections, but not signing out the account.

**Comment by alkisg** [comment]
Yes, to sign them out you would need to do what I wrote above:
https://github.com/ltsp/ltsp/discussions/357#discussioncomment-254025

**Comment by alkisg** [comment]
The tricky part is that we want to notify all previous sessions to gracefully terminate; i.e. the kill signal should be sent to the LTSP clients where the user still has active sessions, not to the LTSP server sftp process. One way to implement this that works over any file system, including SSHFS, is:

1. If files exist under ~/.config/ltsp/logins/, goto step 4, else goto step 2.
2. Create a "session lock" file: `echo HOSTNAME=$(hostname) >~/.config/ltsp/logins/$date`
3. Monitor its existence in the background and terminate all user processes as soon as it's deleted; at the same time, continue running the session: `tail -f ~/.config/ltsp/logins/$date && pkill -u $(id -un) & /usr/sbin/mate-session`
4. Step 4 only happens if there are already other sessions running. Display a warning dialog with the hostnames and dates of all the ~/.config/ltsp/logins/$date files. Ask "If you continue, all those sessions will be disconnected. Are you sure? Yes/Cancel".
5. On "Cancel", just exit. On Yes, remove all the lock files, then goto step 2 to continue with running the session.

By default, SSHFS uses `-o cache_timeout=20`, so all the other sessions will be notified that their lock files were deleted at most 20 seconds later, and all their processes will be killed. 20 seconds is usually fine, but in some fast kiosks where firefox autostarts, one could use SSH_OPTIONS to lower it to 5.

**Comment by alkisg** [comment]
I uploaded a first implementation of a new `ltsp session` applet that for the time being will be undocumented.
New LTSP packages that contain it are available in the proposed PPA.
It can be activated with something like this in ltsp.conf:

```shell
POST_INIT_LTSP_SESSION="sed 's|^Exec=|&/usr/sbin/ltsp session |' -i /usr/share/xsessions/*.desktop"
```

This then shows a "you are currently logged on, disconnect other sessions?" dialog before login.
If the user clicks on the "Yes" button, the other sessions will be automatically disconnected, up to 20 seconds later.

**Reply by perlland** [reply]
Thank You for your considerable time and solution. I ll try and get back to you...Big Thanks

**Reply by macroendrix** [reply]
Tested and works very well, but the mouse cursor doesn't shows when the alert dialog box appears, nevertheless, really good job. Waiting now for documentation. Thanks!

**Reply by shredderique** [reply]
This solution rocks! It does exactly what is needed. There is just one caveat: When I have tested it in an VM environment the logout always occurred after the new login. Wouldn't it be great to specify a certain delay for the new session login e.g. 20 secs?
I don't know what a user would have to do in order to break things within 20 secs when switching to a new session - but in case there is such a possibility ... we have a certain type of users which will definitely figure it out :smiley:

**Comment by intelliant01** [comment]
@alkisg Found a caveat - If a user reboots without logging out from the session, then the session lock file remains and gives impression of a false positive upon next login attempt.

**Comment by alkisg** [comment]
Check `ps faux` on the server, I believe you'll see that the login/systemd etc processes of the user are still running.
It's not a lock file but running processes; LTSP isn't involved in their termination.

**Comment by intelliant01** [comment]
I had already checked `ps faux` and there are no processes of the user. 

Have reproduced it consistently and confirmed with x86 clients booting a Ubuntu 20.04 image by doing a `watch ls -l /home/<user>/.config/ltsp/logins/` and found that the file `~/.config/ltsp/logins/$date` remains if a user does a restart/shutdown instead of logging out. 

This is not the case with a RPi 400 or RPi 4B booting an ARM64 Ubuntu 20.04 image. That works exactly as expected.

Screenshot from server depicting status shortly after the user 'teacher1' does a shutdown/reboot ~~logs out~~ -

![Screenshot at 2022-03-08 15-02-04](https://user-images.githubusercontent.com/2577060/157209186-b2028d1f-a070-4951-9eb2-595c7d50a854.png)

If there would have been login/systemd processes of the user then would it still show the custom dialog? I guess not.

**Comment by alkisg** [comment]
That would mean that the client OS doesn't cleanly terminate the user processes on reboot.
Which one has the issue, Ubuntu MATE 20.04?

**Comment by intelliant01** [comment]
Yes, Ubuntu MATE 20.04 on a x86 client only. Ubuntu MATE 20.04 arm64 client works fine.

**Comment by alkisg** [comment]
It's probably related to shutdown speed, i.e. arm64 could have the same issue if rpi400 was faster.
Anyway this whole session-warning is experimental, patches welcome but I don't want to spend too much time on it personally.
Maybe some time in the future when ltsp grows an https service, that one can be reused to properly track clients etc.

**Reply by intelliant01** [reply]
Not a deal breaker at all. It is just a nice-to-have as of now. :)
